phabricatoremails/worker.py

```diff
@@ -0,0 +1,111 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import signal
+import time
+from dataclasses import dataclass
+from logging import Logger
+from typing import Callable
+
+from phabricatoremails.db import DB
+from phabricatoremails.query_position_store import QueryPositionStore
+from phabricatoremails.thread_store import ThreadStore
+
+PIPELINE_CALLBACK = Callable[[ThreadStore, int], int]
+
+
+@dataclass
+class PhabricatorWorker:
+    """Runs the email pipeline continuously until service shutdown
+
+    The position on the Phabricator feed is stored in persistence, and the time between
+    polling the feed is configurable.
+
+    Continually runs until the process is stopped. To guard against errors or the
+    process being killed mid-pipeline, each loop is guarded in a database transaction.
+    When the process is killed, any progress on the current batch of events is
+    rolled back.
+    """
+
+    _logger: Logger
+    _poll_gap_seconds: int
+
+    @staticmethod
+    def _poll(
+        query_position_store: QueryPositionStore,
+        thread_store: ThreadStore,
+        pipeline: PIPELINE_CALLBACK,
+    ):
+        query_position = query_position_store.get_position()
+        last_key = pipeline(thread_store, query_position.up_to_key)
+
+        if query_position.up_to_key != last_key:
+            query_position.up_to_key = last_key
+            return False
+        else:
+            return True
+
+    def process(self, db: DB, pipeline: PIPELINE_CALLBACK):
+        shutdown_requested = False
+        is_sleeping = False
+
+        def on_shutdown_signal(_, __):
```

Comment:
Since the parameters are unused, I'd like to avoid setting a name.
Using `_` correctly implies that the parameter isn't used below.

I'll update this to:
```
def on_shutdown_signal(*_):
    ...
```